jetty-core/jetty-security/src/main/java/org/eclipse/jetty/security/SecurityHandler.java

```diff
@@ -0,0 +1,769 @@
+            matches.sort(this::compare);
+            if (LOG.isDebugEnabled())
+                LOG.debug("getConstraint {} -> {}", pathInContext, matches);
+            Constraint constraint = null;
+            for (MappedResource<Constraint> c : matches)
+                constraint = Constraint.combine(constraint, c.getResource());
+
+            return constraint;
+        }
+
+        /**
+         * {@link Comparator} method to sort paths from least specific to most specific. Using
+         * the {@link #pathSpecGroupPrecedence(PathSpecGroup)} to rank different groups and
+         * {@link PathSpec#getSpecLength()} to rank within a group.  This method may be overridden
+         * to provide different precedence between constraints.
+         * @param ps1 the first {@code PathSpec} to be compared.
+         * @param ps2 the second {@code PathSpec} to be compared.
+         * @return -1, 0 or 1
+         */
+        @Override
+        public int compare(PathSpec ps1, PathSpec ps2)
+        {
+            PathSpecGroup g1 = ps1.getGroup();
+            PathSpecGroup g2 = ps2.getGroup();
+            if (g1.equals(g2))
+                return Integer.compare(ps1.getSpecLength(), ps2.getSpecLength());
+
+            return Integer.compare(pathSpecGroupPrecedence(g1), pathSpecGroupPrecedence(g2));
+        }
+
+        int compare(MappedResource<Constraint> c1, MappedResource<Constraint> c2)
+        {
+            PathSpecGroup g1 = c1.getPathSpec().getGroup();
+            PathSpecGroup g2 = c2.getPathSpec().getGroup();
+            int l1 = c1.getPathSpec().getSpecLength();
+            int l2 = c2.getPathSpec().getSpecLength();
+
+            if (g1.equals(g2))
+                return Integer.compare(l1, l2);
+
+            return Integer.compare(pathSpecGroupPrecedence(g1), pathSpecGroupPrecedence(g2));
+        }
+
+        /**
+         * Get the relative precedence of a {@link PathSpecGroup} used by {@link #compare(MappedResource, MappedResource)}
+         * to sort {@link Constraint}s.  The precedence from most significant to least is:
+         * <ul>
+         *     <li>{@link PathSpecGroup#EXACT}</li>
+         *     <li>{@link PathSpecGroup#ROOT}</li>
+         *     <li>{@link PathSpecGroup#SUFFIX_GLOB}</li>
+         *     <li>{@link PathSpecGroup#MIDDLE_GLOB}</li>
+         *     <li>{@link PathSpecGroup#PREFIX_GLOB}</li>
+         *     <li>{@link PathSpecGroup#DEFAULT}</li>
+         * </ul>
+         * @param group The group to rank.
+         * @return An integer representing relative precedence between {@link PathSpecGroup}s.
+         */
+        protected int pathSpecGroupPrecedence(PathSpecGroup group)
+        {
+            return switch (group)
+            {
+                case EXACT -> 5;
+                case ROOT -> 4;
+                case SUFFIX_GLOB -> 3;
+                case MIDDLE_GLOB -> 2;
+                case PREFIX_GLOB -> 1;
+                case DEFAULT -> 0;
```

Comment:
Should not `PathSpecGroup` implement `Comparable` instead?
You would do `g1.compareTo(g2)` where necessary.

Reply:
It does, but it implements the servlet spec order.  Besides, doing it this way makes it extensible in specialized securityhandlers